### ISSUE TITLE
Data files in /data package directories not rebuilt when source data changed

### ISSUE THREAD

**Issue by jrawbits** [VisionEval/VE-Installer#13]
VisionEval modules often contain .rda files (R data) in the package/data subdirectory.  These can be rebuilt manually but are apparently not rebuilt during the 'make' operation that rebuilds the packages and assembles the runtime VisionEval installation. Superficially, it appears that when changes are made to the source .csv files and the 'make' operation is performed to rebuild the packages,, the .rda files are not rebuilt even though other code does appear to be.

**Comment by jrawbits** [comment]
In fact, the packages do appear to get rebuilt correctly with the updated data by VE-Installer.  If a rebuilt package containing data is loaded, and the data() function is used to retrieve one of its datasets, the changed data appears to be present.

It is true that the .rda files are not rebuilt in the file system.  But the .rda files themselves are unnecessary - they will always be rebuilt when the package is built (as a Windows binary) or installed (from source).  In fact, they probably should never have been versioned in Github in the first place.

That said, it may be useful when debugging changes to the data generation code to generate the .Rda files so they can be inspected, which may be accomplished from within RStudio by running devtools::build with "binary=TRUE" or devtools::install or devtools::check - the .Rda files are regenerated as a side effect of the build (VE-Installer hides these artifacts in a temporary directory and then disposes of them later).  Of course, it is also possible just to run the package R code manually to have it rebuild the files.

I'm going to leave this issue open until others confirm that the updated data is, in fact, being placed into the installed package.

**Comment by jrawbits** [comment]
@arashasadabadi  A couple of questions about the workaround for rebuilding the /data contents (using devtools::check to manually build an installable package):  does devtools::build (rather than devtools::check) rebuild the data successfully? What is the reproducible set of steps that shows that the data is not being built by VE-Installer, but is being rebuilt by manual devtools::check (e.g. set up a controlled modification, run the rebuild/install using the two approaches, run lines of R code to show that the two package build approaches generate different numbers).

A short script to run in the original installed VE, followed by rebuilding the changed package, followed by running the script again would do the trick. Then I can play with putting packages generated in the various ways in different libraries and examine how it is not getting rebuilt.

**Comment by arashasadabadi** [comment]
@jrawbits No devtool::build also does not update the data.

I tried to write a memo for my own user experience for a speicif data I wanted to change with all these three methods( devtools::install, devtools::install, devtools::check) and attached it here.

You can follow the same path and see if you can reproduce the same results.
[memo_update_pakages_data.docx](https://github.com/VisionEval/VE-Installer/files/3340550/memo_update_pakages_data.docx)

**Comment by jrawbits** [comment]
The problem is this: the converstion of "/inst/extdata" to built R data objects in the "/data" folder only happens when a binary build takes place (as part of running all the R code for the package).  However when the "/data" directory is created that way, the package build process sees it as a disposable artifact and does not actually add it into a package - what ends up in /data is what was there at the beginning of the build, not what is there at the end of the build.

Running "devtools::check" manually to build the package evades that problem because devtools::check apparently builds the package twice (once to look for errors, and once to see if the result is installable).

So fixing this problem (which is important) will require surgery on the framework and packages so we can slip in an extra build step to recreate /data BEFORE any package building starts.

If we set up some standard callable approach for each module to rebuild its /data, we could automate that into the process.

I'm researching whether there is some recommended way in 'devtools' land to automate building /data from /inst/extdata.

**Comment by arashasadabadi** [comment]
@jrawbits Thanks for looking into this. For the first proposed solution, a short term solution could be using devtools:check in the installer script for each package before using ::install function.

**Comment by jrawbits** [comment]
I may take that route in the short run, but it would entail building the package (and the data) over and over and over - it gets rebuilt again (but not used) when we create the binary package (or install the source).

There ultimately needs to be a much cleaner separation between estimating the required data sets and building the runtime packages. Unfortunately, it looks like that may require some deep reorganization of the R scripts to distinguish the setup that builds the data from the setup that constructs the module functions.

**Comment by arashasadabadi** [comment]
I agree, once we want to take that route seriously I will comment more on the issue (52) you opened on dev branch. but as of now, I know it will need a quiet load of manual work of code splitting and error checking (don't have an exact estimate but it's a lot of work).
